File: Javascript/Structural Patterns/Proxy/proxy-exII.ts
export interface Service {
    request(user: string): void
}

class RealService implements Service {
    request(user: string): void {
        console.log(`RealService: Handle ${user} request`)
    }
}

export class ProxyService implements Service {
    private realService: RealService
    private allowUsers: Set<string>

    constructor() {
        this.realService = new RealService()
        this.allowUsers = new Set(["admin", "user1"])
    }

    request(user: string): void {
       if (!this.allowUsers.has(user)) {
            console.log(`ProxyService: Reject ${user} request`)
        }
       this.realService.request(user)
    }
}

File: Go/Structural_Patterns/Proxy/proxy-exII.go
package proxy

import "fmt"

type Service interface {
	Request(user string)
}

type RealService struct{}

func (r *RealService) Request(user string) {
	fmt.Printf("ProxyService: Receive %s request\n", user)
	fmt.Printf("RealService: Handle %s request\n", user)
}

type ProxyService struct {
	realservice *RealService
	allowUsers  map[string]bool
}

func (p *ProxyService) Request(user string) {
	if !p.allowUsers[user] {
		fmt.Printf("ProxyService: Reject %s request\n", user)
		return
	}
	p.realservice.Request(user)
}

func NewProxyService() *ProxyService {
	return &ProxyService{
		realservice: &RealService{},
		allowUsers: map[string]bool{
			"admin": true,
			"user1": true,
		},
	}
}
